src/main/java/com/example/servlet20/AuthenticationFilter.java

```diff
@@ -0,0 +1,47 @@
+package com.example.servlet20;
+
+import jakarta.servlet.*;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Enumeration;
+
+public class AuthenticationFilter implements Filter {
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        Filter.super.init(filterConfig);
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
+        String basicAuthCode = "";
+        final Enumeration<String> headerNames = ((HttpServletRequest) request).getHeaderNames();
+        boolean authDetailsFound = false;
+        while (headerNames.hasMoreElements()) {
+            final var headerName = headerNames.nextElement();
+            if (headerName.equalsIgnoreCase("authorization")) {
+                authDetailsFound = true;
+                basicAuthCode = ((HttpServletRequest) request).getHeader(headerName);
+            }
+        }
+        if(!authDetailsFound) {
+            response.getWriter().println("You must enter authentication details!");
```

Comment:
within if clause add: 1) setting 401 2) return statement. Because it won't make sense execute the rest of the code if no auth header is found